rfcs/2020-04-15-wasm-foreign-modules.md

```diff
@@ -0,0 +1,559 @@
+* In the future, support something like AssemblyScript as a first class language.
+
+
+## Drawbacks
+
+WASM is still **very young**. We will be early adopters and will pay a high adoption price. Additionally, we may find
+ourselves using unstable functionality of some crates, including a non-trivial amount of `unsafe`, which may lead to
+unforseen consequences.
+
+Our **binary size will bloom**. We currently (in unoptimized POC implementations) see our binary size grow by over 60
+MB. The Lucet team is aware of this issue, and in some preliminary discussions this binary size increase is unexpected and
+we expect it to improve. If this does not improve, we can consider adopting `wasmtime`, lucet's sister JIT implementation
+which has a smaller binary footprint.
+
+Our team will need to diagnose, support, and evolve an arguably rather **complex API** for WASM modules to work
+correctly. This will mean we'll have to think of our system very abstractly, and has many questions around API stability
+and compatibility.
+
+This feature is **advanced and may invoke user confusion**. Novice or casual users may not fully grasp the subtleties
+and limitations of WASM. We must practice caution around the user experience of this feature to ensure novices and
+advanced users can understand what is happening in a Vector pipeline utilizing WASM modules.
+
+
+## Outstanding Questions
+
+Since this is a broad reaching feature with a number of green sky ideas, we should discuss these questions:
+
+
+### Agree on breadth and depth of implementation
+
+We should support modules so broadly?
+
+* As sinks? For this we should consider how we handle batching, partitioning, etc, more.
+  * **Temporary Conclusion:** Held off on deciding. Some interest.
+* As sources? This requires thinking about how we can do event sourcing.
+  * **Temporary Conclusion:** Held off on deciding. Some interest.
+* What about the idea of codecs being separate?
+  * **Temporary Conclusion:** Held off on deciding. Some interest.
+* This RFC makes some provisions for a future Event refactor, is it possible that this might happen?
+  * **Temporary Conclusion:** Held off on deciding. Some interest.
+
+
+### Consider supporting a blessed compiler
+
+Our original discussions included the idea of a blessed compiler and a UX similar to our existing `lua` or `javascript`
+transform. Indeed, our initial implementation should include just a transform, as it's by far the easiest place to
+introduce this feature.
+
+During investigation of the initial POC we determined that our most desirable "blessed" language would be
+AssemblyScript, unfortunately, it's currently not able to run inside Lucet. Work is ongoing and we expect it to be
+possible soon.
+
+**Temporary Conclusion:** AssemblyScript and other languages we evaluated did not provide a way to package themselves as
+WASM modules, making it hard to integrate with them. We are planning to revisit this at a later date when the ecosystem
+is more mature.
```

Comment:
I think this should be considered critical for v1. I doubt many users would be thrilled over the prospect of writing plugins in c, and writing them in rust means we're competing against simply forking and writing components natively, which given it has benefits in terms of portability might end up winning for some use cases.

I think the teams that need wasm plugins the most are also going to need a more approachable language.

Reply:
So folks can always write in assemblyscript outside vector and build it into a `.wasm` then give it to vector. We just can't support inlined assemblyscript etc. So we can't support this immediately:

```toml
[transforms.foo]
type = "assemblyscript"
inline = """
foo = "bar"
"""
```

That'll need more work and be part of a later version.

Reply:
I think this is fine for v1.